Add a guide page on credential types (#33745)
mdn/content: files/en-us/web/api/credential_management_api/credential_types/index.md

---
title: Credential types
slug: Web/API/Credential_Management_API/Credential_types
page-type: guide
---

{{DefaultAPISidebar("Credential Management API")}}

The Credential Management API enables a website to create, store, and retrieve {{glossary("credential", "credentials")}}, to help a user securely log into websites. Its supports four different types of credential, which are all represented as subclasses of the {{domxref("Credential")}} interface:

> - **Passwords** ({{domxref("PasswordCredential")}})

Author: Hm, because it makes it look like a method call you mean? What I really want to say is "represented by" but thought that might be a bit too repetitive.

Do you like a table for this -> https://github.com/mdn/content/pull/33745/commits/13cde577aa3dd3ca5a151a8f53db236c7e5b2d68 ?

> - **Passwords** ({{domxref("PasswordCredential")}})

Reviewer: Yes, I think that is clearer; nice one @wbamberg

![Class diagram showing the five different credential subclasses.](credential-types.svg)

In this guide we'll introduce the different credential types and explain at a high level how they are used.

> **Note:** Although we're describing all the credential types together here, the different credential types are defined in various different specifications, which extend the main Credential Management API specification.
>
> - [Credential Management API](https://w3c.github.io/webappsec-credential-management/) defines passwords and legacy federated credentials.
> - [Federated Credential Management API](https://fedidcg.github.io/FedCM/) defines the new federated credentials.
> - [WebOTP API](https://wicg.github.io/web-otp/) defines OTP credentials.
> - [Web Authentication API](https://w3c.github.io/webauthn/) defines Web Authentication assertions.

## Passwords

> **Note:** Most browsers do not support this credential type and it is not widely used on the web. Instead, browsers automatically offer to store passwords in the password manager, and can automatically retrieve stored passwords to autofill [password input elements](/en-US/docs/Web/HTML/Element/input/password).

> Modern browsers provide users with a password manager, which enables users to store the passwords they enter on websites, and later retrieve them when they need to log in again. Password managers can help with password security by remembering passwords for users and autofilling them, which allows users to choose stronger passwords.

Author: I think it's overall better practice to have strong unique passwords and write them down somewhere than to have simple enough passwords that they can be remembered: https://www.schneier.com/blog/archives/2005/06/write_down_your.html.

> Modern browsers provide users with a password manager, which enables users to store the passwords they enter on websites, and later retrieve them when they need to log in again. Password managers can help with password security by remembering passwords for users and autofilling them, which allows users to choose stronger passwords.

Reviewer: I agree with this, but even so, it is better for security to not have to write them down at all, regardless of how strong or weak they are.

That said, I will leave it up to you whether you add it. I don't think it is that important to the narrative that I'm going to die on this hill ;-)

![Sequence diagram showing sign-in with a password credential.](password-get.svg)

## Federated identity credentials

In a {{glossary("federated identity")}} system, a separate entity acts as an intermediary between the user and the website which they are trying to sign into. This entity, called an {{glossary("identity provider")}} (IdP), manages the user's credentals and can authenticate users, and is trusted by the website to make assertions about a user's identity.

The user has an account with the IdP: when they need to sign into the website they authenticate with the IdP. The IdP then returns a token to the user's browser, which the browser delivers to the website. The website verifies the token and if verification succeeds, signs the user in.

Federated identity is often provided as a service by corporations: for example, users who have Google, Microsoft, or Facebook accounts can use them to sign into websites that support them.

> Support for federated identity in the Credential Management API was originally provided through the {{domxref("FederatedCredential")}} interface. However, this mechanism depends on technologies such as third-party cookies, which are intrinsically privacy-invasive, and as these technologies have been deprecated by browsers, new approaches are needed.

Reviewer: ```suggestion
Support for federated identity in the Credential Management API was originally provided through the {{domxref("FederatedCredential")}} interface. However, this mechanism depends on technologies such as [third-party cookies](/en-US/docs/Web/Privacy/Third-party_cookies), which are intrinsically privacy-invasive. These technologies were [deprecated in browsers](/en-US/blog/goodbye-third-party-cookies/), therefore a new approach was needed.
```

> The [Federated Credential Management API](/en-US/docs/Web/API/FedCM_API) defines a privacy-preserving mechanism for federated identity on the web. In this mechanism, you call {{domxref("CredentialsContainer.get", "navigator.credentials.get()")}} to request a federated identity credential, and this triggers a protocol exchange between the browser and the IdP.

Reviewer: ```suggestion
The [Federated Credential Management API](/en-US/docs/Web/API/FedCM_API) defines a privacy-preserving mechanism for federated identity on the web. You start by calling {{domxref("CredentialsContainer.get", "navigator.credentials.get()")}} to request a federated identity credential, and this triggers a protocol exchange between the browser and the IdP.
```

If, in the course of this exchange, the user can be authenticated with the IdP, the browser returns an {{domxref("IdentityCredential")}} object in the fulfilment of the `Promise` returned from `get()`. The website front end code can send this to the server for verification.

![Sequence diagram showing sign-in using a federated identity credential.](fed-cm-get.svg)

Note that {{domxref("CredentialsContainer.create", "create()")}} and {{domxref("CredentialsContainer.store", "store()")}} are not used when working with the Federated Credential Management API.

## One-time passwords

> A one-time password (OTP) is an authentication technique in which the website has a means to send messages to a user (such as an email address or a phone number) and uses this to send a unique code to the user, who then must enter the code on the site to prove their control of the communications endpoint. Websites sometimes use this as a second authentication factor in addition to a password.

Reviewer: ```suggestion
A one-time password (OTP) is an authentication technique in which the website sends a unique code to the user via a messaging system such as email or SMS. The user must then enter the code on the site to prove their control of the communications endpoint. Websites sometimes use this as a second authentication factor in addition to a password.
```

The [WebOTP API](/en-US/docs/Web/API/WebOTP_API) defines the {{domxref("OTPCredential")}} interface, which solves a specific usability problem in this exchange: when a user receives the code, they have to open a different application, find the message, then copy the code into a form on the website. This is awkward, especially on a mobile device, and especially when the device receiving the message is the same as the device being used to sign into the site.

In browsers that support the `OTPCredential` type, a website can call {{domxref("CredentialsContainer.get", "navigator.credentials.get()")}}, asking for an OTP credential, then ask the server to send the message containing the code (only SMS is supported as a transport). The server must send a specially formatted SMS message, which the browser can read.

The browser then returns an `OTPCredential` object in the fulfilment of the `Promise` returned from `get()`, and this object contains the code. The website front end can then use the code to autofill an input element on the site, or submit the code to the server automatically.

> ![Sequence diagram showing sign-in using an OTP credential.](otp-get.svg)

Reviewer: I'm not sure about the third line down on this diagram — the one that says "Send SMS". The website doesn't send an SMS to the server — it sends a request to the server to send an SMS to the client.

So, perhaps the third line down should say something like "Request server to send SMS", and the fourth line down should be "Send SMS"?

Author: That's fair. I suppose I wanted to avoid having a fifth entity here, so "send SMS" is a kind of shorthand for "ask the server to send an SMS".

Author: I've updated the diagram in https://github.com/mdn/content/pull/33745/commits/1fabf7453f3b4691193f3eaf84dbbd650338ec8c.

Reviewer 2: ![image](https://github.com/mdn/content/assets/5368500/4f02e999-e567-4528-8661-3c7fb3cfe56f)

So just to be clear on this - the website backend generates a code and sends it to the SMS server _via SMS" - that's what you mean by "then ask the server to send the message containing the code (**only SMS is supported as a transport**)"? 

I guess this allows for other transports in future, but it does feel odd that the website is capable of sending an SMS code to the user, but does so via an intermediary.

This is a little confusing. Suggestion, change image so that instead of `send SMS(code)` it is `send code (SMS transport)` or similar.

Author: > So just to be clear on this - the website backend generates a code and sends it to the SMS server _via SMS" 

No. The website backend generates a code then _causes an SMS containing that code to be sent to the device_. In fact how it does that is really an implementation detail in the website backend, IMO. I mean it could do it by having a guy text it from their cellphone, AFAIK, and it would work, just not be very scalable.

I added "SMS server" in response to Chris's comment:

> I'm not sure about the third line down on this diagram — the one that says "Send SMS". The website doesn't send an SMS to the server — it sends a request to the server to send an SMS to the client.
>
> So, perhaps the third line down should say something like "Request server to send SMS", and the fourth line down should be "Send SMS"?

on the previous version of the diagram, that looked like this:

<img width="847" alt="Screen Shot 2024-05-30 at 7 55 56 PM" src="https://github.com/mdn/content/assets/432915/44254235-648b-4b36-a600-3717399c7c20">

...but perhaps it was clearer before?

Reviewer 2: My take is that "Website back end" implies any servers that it uses, including any SMS server. 

That said I don't have a problem with having the server line. The problem is that `send SMS(code)` implied to me that the backend sent an SMS, and this was reinforced by the text "ask the server to send the message containing the code". 

You can fix this by changing this sentence to add "SMS:

> ask the **SMS** server to send the message containing the code

And update image from `send SMS(code)` to `request SMS code`.

Or revert.

But let's see what @chrisdavidmills says

Author: > this was reinforced by the text "ask the server to send the message containing the code".

Yeah, that was because I updated the diagram but forgot to update the text.

I agree that either updating the text or reverting the diagram should work, and am happy to see what Chris thinks.

Reviewer: I think @hamishwillee's suggestion works well:

> You can fix this by changing this sentence to add "SMS:
>
>>    ask the SMS server to send the message containing the code
>
> And update image from send SMS(code) to request SMS code.

I don't think we need an additional line for "SMS server"; that just overcomplicated it. I too, am taking "Website back end" to mean "any and all servers that make up the back end portion of the application, which could include an SMS server".

Author: -> https://github.com/mdn/content/pull/33745/commits/bc7a29b3371a4c6fc603534e1dd3efadf18ffef8

---
title: Credential types
slug: Web/API/Credential_Management_API/Credential_types
page-type: guide
---

{{DefaultAPISidebar("Credential Management API")}}

> The Credential Management API enables a website to create, store, and retrieve {{glossary("credential", "credentials")}}, to help a user securely log into websites. It supports four different types of credentials:

Reviewer 2: ```suggestion
The Credential Management API enables a website to create, store, and retrieve the {{glossary("credential", "credentials")}} that allow a user to securely log in. It supports four different types of credentials:
```

| Type                    | Interface                                                                          |
| ----------------------- | ---------------------------------------------------------------------------------- |
| Password                | {{domxref("PasswordCredential")}}                                                  |
| Federated identity      | {{domxref("FederatedCredential")}} (deprecated), {{domxref("IdentityCredential")}} |
| One-time password (OTP) | {{domxref("OTPCredential")}}                                                       |
| Web Authentication      | {{domxref("PublicKeyCredential")}}                                                 |

The credential types are all represented as subclasses of the {{domxref("Credential")}} interface:

> ![Class diagram showing the five different credential subclasses.](credential-types.svg)

Reviewer 2: Ditto for diagram - if it is deprecated have it below identiy cred and perhaps put box in grey?

Author: -> https://github.com/mdn/content/pull/33745/commits/bf06e0b52faa11f2089e4d18ca5c6ebe029b5131

In this guide we'll introduce the different credential types and explain at a high level how they are used.

> > **Note:** Although we're describing all the credential types together here, the different credential types are defined in various different specifications, which extend the main Credential Management API specification.

Reviewer 2: ```suggestion
> **Note:** Although we're describing all the credential types together here, the different credential types are defined in a number of different specifications, which extend the main Credential Management API specification.
```

Reviewer: "multiple", or "several"? "a number of" always sounds a bit vague and scrappy to me.

>
> - [Credential Management API](https://w3c.github.io/webappsec-credential-management/) defines passwords and legacy federated credentials.
> - [Federated Credential Management API](https://fedidcg.github.io/FedCM/) defines the new federated credentials.
> - [WebOTP API](https://wicg.github.io/web-otp/) defines OTP credentials.
> - [Web Authentication API](https://w3c.github.io/webauthn/) defines Web Authentication assertions.

## Passwords

> > **Note:** Most browsers do not support this credential type and it is not widely used on the web. Instead, browsers automatically offer to store passwords in a password manager, and can automatically retrieve stored passwords to autofill [password input elements](/en-US/docs/Web/HTML/Element/input/password).

Reviewer 2: So just to be clear the browser automatically stores credential types when needed/requested, and autofills forms or whatever, while the credential API just stores and gets credentials, but all the work for doing this and for populating forms has to be done by the developer?

If so, I wonder why any browser/site has implemented this. 

Perhaps make first _automatically_ italic.

Author: Well I'm not sure, I only learned last week or so that this API isn't really used, because neither MDN [nor web.dev](https://web.dev/articles/security-credential-management) say anything about that. One thing you can do with CM, that the browser won't do for you, is sign people in automatically. I suppose in general gives you much more control over how sign-in happens.

Reviewer 2: As it is now the whole rationale is a bit wishy washy, but I'd say this is more than good enough for the current PR. We can find out more in a follow on.

![Sequence diagram showing sign-in with a password credential.](password-get.svg)

## Federated identity credentials

In a {{glossary("federated identity")}} system, a separate entity acts as an intermediary between the user and the website they are trying to sign into. This entity, called an {{glossary("identity provider")}} (IdP), manages the user's credentials, can authenticate users, and is trusted by the website to make assertions about a user's identity.

The user has an account with the IdP: when they need to sign into the website they authenticate with the IdP. The IdP then returns a token to the user's browser, which the browser delivers to the website. The website verifies the token and, if verification succeeds, signs the user in.

Federated identity is often provided as a service by corporations: for example, users who have Google, Microsoft, or Facebook accounts can use them to sign into websites that support them.

> Support for federated identity in the Credential Management API was originally provided through the {{domxref("FederatedCredential")}} interface. However, this mechanism depends on technologies such as [third-party cookies](/en-US/docs/Web/Privacy/Third-party_cookies), which are intrinsically privacy-invasive. These technologies were [deprecated in browsers](/en-US/blog/goodbye-third-party-cookies/), therefore a new approach was needed.

Author: -> https://github.com/mdn/content/pull/33745/commits/b25f13dddb907a0e8688e6b53dbad3ab96c8b15c

## One-time passwords

A one-time password (OTP) is an authentication technique in which the website sends a unique code to the user via a messaging system such as email or SMS. The user must then enter the code on the site to prove their control of the communications endpoint. Websites sometimes use this as a second authentication factor in addition to a password.

The [WebOTP API](/en-US/docs/Web/API/WebOTP_API) defines the {{domxref("OTPCredential")}} interface, which solves a specific usability problem in this exchange: when a user receives the code, they have to open a different application, find the message, then copy the code into a form on the website. This is awkward, especially on a mobile device, and especially when the device receiving the message is the same as the device being used to sign into the site.

In browsers that support the `OTPCredential` type, a website can call {{domxref("CredentialsContainer.get", "navigator.credentials.get()")}}, asking for an OTP credential, then ask the server to send the message containing the code (only SMS is supported as a transport). The server must send a specially formatted SMS message, which the browser can read.

> The browser then returns an `OTPCredential` object in the fulfilment of the `Promise` returned from `get()`, and this object contains the code. The website front end can then use the code to autofill an input element on the site, or submit the code to the server automatically.

Reviewer 2: ```suggestion
The browser then returns an `OTPCredential` object in the fulfilment of the `Promise` returned from `get()`, and this object contains the code. The website front end can use the code to autofill an input element on the site, or submit the code to the server automatically.
```

![Sequence diagram showing sign-in using an OTP credential.](otp-get.svg)

Note that {{domxref("CredentialsContainer.create", "create()")}} and {{domxref("CredentialsContainer.store", "store()")}} are not used when working with OTP credentials.

## Web Authentication assertions

The [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API) (WebAuthn) enables users to log into websites by asking an _authenticator_ to generate digitally signed assertions about a user's identity.

> An authenticator is an entity that is inside or attached to the user's device, and that can perform the cryptographic operations needed to register and authenticate users, and securely store the cryptographic keys used in these operations. An authenticator might be integrated into the device, like the [Touch ID](https://en.wikipedia.org/wiki/Touch_ID) system in Apple devices or the [Windows Hello](https://en.wikipedia.org/wiki/Windows_10#System_security) system, or it might be a removable module like a [Yubikey](https://en.wikipedia.org/wiki/YubiKey).

Reviewer 2: FMI could this be the google authenticator app, or similar? Just wondering where that kind of OTP might fit in.
